src/ILCompiler.Compiler/src/Compiler/NameMangler.cs

```diff
@@ -76,8 +76,35 @@ internal string SanitizeName(string s, bool typeName = false)
             return (sb != null) ? sb.ToString() : s;
         }
 
+        /// <summary>
+        /// Dictionary given a mangled name for a given <see cref="TypeDesc"/> 
+        /// </summary>
         private ImmutableDictionary<TypeDesc, string> _mangledTypeNames = ImmutableDictionary<TypeDesc, string>.Empty;
 
+        /// <summary>
+        /// Set of names already computed for <see cref="_mangledTypeNames"/>.
+        /// Needed to ensure bijective mapping between a <see cref="TypeDesc"/> and a mangled name.
+        /// </summary>
+        private ImmutableHashSet<string>_reversedMangledTypeNames = ImmutableHashSet<string>.Empty;
```

Comment:
This should not be needed to be kept around because of assembly simple name is always prepended in front of type name.

A local temporary set that just lives while the names within given assembly are getting generated should be sufficient.